PYIC-5875: Test harness for API tests

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Test harness for API tests

### Why did it change

This adds the basics of a new cucumber testing project for running API tests through core. It includes one feature that runs a happy path journey through an app journey.

It includes a GHA workflow to run the tests pre-merge.

Lots of details in the commit messages for anyone interested.

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-5875](https://govukverify.atlassian.net/browse/PYIC-5875)


[PYIC-5875]: https://govukverify.atlassian.net/browse/PYIC-5875?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Wynndow committed:
PYIC-5875: Initial commit for api tests framework

This creates a node project using node 22 - appears to be the current
and going to LTS in October.

Includes typescript with the defaults for the tsconfig file as provided
by running `npx tsc --init`

Wynndow committed:
PYIC-5875: Set TS target to es2023

It's the latest and seems to be supported since node v18

Wynndow committed:
PYIC-5875: Set type to module in packaage.json

This makes the modeul type ECMAScipt instead of CommonJS. It's more
modern.

Wynndow committed:
PYIC-5875: Set tsconfig/module to nodenext

It reflects the module system for node16 (or later) which supports ES
modules and CJS modules.

Wynndow committed:
PYIC-5875: Install cucumber

Using npm install --save-dev @cucumber/cucumber

Wynndow committed:
PYIC-5875: Add running cucumber as a script

This runs a binary installed by the cucumber package:
https://github.com/cucumber/cucumber-js/blob/HEAD/docs/cli.md#cli

Wynndow committed:
PYIC-5875: Add cucumber config file

With some basic options. All the options available can be found here:
https://github.com/cucumber/cucumber-js/blob/HEAD/docs/configuration.md

Wynndow committed:
PYIC-5875: Install ts-node

It's an execution engine for TS code. It'll transpile it JIT, so we
won't have to pre-compile everything before running tests. It's
recommended in the cucumber docs.

Wynndow committed:
PYIC-5875: Set loader and import cucumber config values

These values are to integrate cucumber with ts-node.

https://github.com/cucumber/cucumber-js/blob/4c67f49687e8039967c217e129454362e4c536ca/docs/transpiling.md#loader-option

Wynndow committed:
PYIC-5875: Downgrade tsconfig.target to es2022

Seems that ts-node can't handle es2023. Running the cucumber tests with
es2023 results in:
```
error TS6046: Argument for '--target' option must be: 'es5', 'es6', 'es2015', 'es2016', 'es2017', 'es2018', 'es2019', 'es2020', 'es2021', 'es2022', 'esnext'.
```

Wynndow committed:
PYIC-5875: Downgrade to node 20

ts-node was producing some deprecation warning with node 22. Also node
20 is the current lts.

Wynndow committed:
PYIC-5875: Increase global timeout to 15s

The default timeout is 5000ms. This can be a bit too short when calling
an api endpoint that triggers a lambda that needs to coldstart.

f

Wynndow committed:
PYIC-5875: Add cucumber-report to gitignore

No one needs that commiting

Wynndow committed:
PYIC-5975: Remove `main` from package.json

It's to specify the entrypoint for an application if required or
imported. We're not doing any of that, we're just running things with
cucumeber.

Wynndow committed:
PYIC-5875: Rename cucumber script to test

Less typing for my lazy fingers

Wynndow committed:
PYIC-5875: Add cucumber setup with P2 happy path

This adds cucumber-js and associated code to run a happy path journey
using the the various APIs.

Wynndow committed:
PYIC-5875: Install data-vocab package

It's a package published by OneLogin that contains TypeScript types for
the Verifiable Credentials that we can use in our code.

https://github.com/govuk-one-login/data-vocab/pkgs/npm/data-vocab

Wynndow committed:
PYIC-5875: Refactor generation of request bodys

This moves the generation of the request bodys our of the clients and
into their own functions.

Wynndow committed:
PYIC-5875: Refactor JSON data files

This splits the way the data for each scenario supplied to the CRI stub
is handled. Hopefully this makes it more readable and sets us up for a
logical way of maintaing the different bits of data we're going to need
as the test suite expands.

Wynndow committed:
PYIC-5875: Refactor session initialisation step

This refactors the initial step to take a parameter for the journey
type to run. This will make it easier to configure other journeys in
future.

Wynndow committed:
PYIC-5875: Move token exchange body generation

Moves the generation of the body for the token exchange out of the
client and into its own method.

Wynndow committed:
PYIC-5875: Install eslint and fix issues

This adds eslint with recommended config for typescript.

Followed the quickstart quide found here: https://typescript-eslint.io/getting-started

Wynndow committed:
PYIC-5875: Install prettier and apply changes

Adds the prettier tool to ensure consistent formatting of JS code.
Follows the instructions laid out here: https://prettier.io/docs/en/install

Wynndow committed:
PYIC-5875: Rename apiTests -> api-tests

To be consistent with the naming convention used in the root of the
repo.

Wynndow committed:
PYIC-5875: Move cucumber TS files into src

This moves the step definitions and programtic config from the features
dir into the src dir. This keeps all the TS files maintained in one
place, rather than having a not super obvious split.

Wynndow committed:
PYIC-5875: Use DI Vocab classes for user identity

Creates an interface for the user-identity retrieved at the end of the
journey, and uses the interfaces defined in the DI vocab package. This
should give us more confidence that core is producing the right shaped
things.

Wynndow committed:
PYIC-5875: Move config to config dir

This puts all the config in once place, and makes it clear which config
is specifically for cucumber.

Also resets the defaults for the build environment to the correct values

Wynndow committed:
PYIC-5875: Run API tests as a pre-merge check

This adds running the API tests to the pre-merge checks. It will run
them against the build environment.

Wynndow committed:
PYIC-5875: Add example .npmrc file

To install the data-vocab package from GitHub pacakges we need to use a
PAT. We can supply this, scoped to the govuk-one-login scope using a
npmrc template. This adds a template and an ignore for the actual file.
Instructions will be added in a readme.

Wynndow committed:
PYIC-5875: Rename redirect URL config value

Uses URL instead of URI to be consistent with other values.

Wynndow committed:
PYIC-5875: Add a README for api tests

Wynndow committed:
PYIC-5875: Remove resourceId from CRI stub calls

The CRI stub has been refactored to handle the resourceId internally.
This means we can simplify the API call.

Joe-Edwards-GDS commented on api-tests/src/config/config.ts:
As discussed - I think it would be preferable to set up a `.env` file for this, which will make it easier to manage configurations as well as making it independent from other setup.

You can use https://www.npmjs.com/package/dotenv to load it programmatically when initialising the tests. I'd suggest a similar pattern of having a `.env.template` and `.env`.

The template can also replace all the defaults in `constants.ts`, and properly separate the config from code. 😄

These support comments, so you can do things like:
```
# Build configuration
SOME_VALUE=build-value
SOME_OTHER_VALUE=boop

# Dev configuration
# SOME_VALUE=dev-value
# SOME_OTHER_VALUE=beep
```

Joe-Edwards-GDS commented on api-tests/package.json:
```diff
@@ -0,0 +1,23 @@
+  "type": "module",
+  "scripts": {
+    "test": "cucumber-js",
+    "lint": "eslint . && prettier . --check",
```
I get style issues on the cucumber html report when I run this - probably worth excluding that if we can!

Wynndow replied:
That's weird, I don't. And I remember when I set it up that it's supposed to ignore stuff in .gitignore. Is eslint or prettier failing?

Joe-Edwards-GDS replied:
```
$ npm run lint

> apitests@1.0.0 lint
> eslint . && prettier . --check

Checking formatting...
[warn] cucumber-report.html
[warn] Code style issues found in the above file. Run Prettier with --write to fix.
```
Looks like prettier

Wynndow replied:
And your .gitignore contains `api-tests/cucumber-report.html`?

Joe-Edwards-GDS replied:
Yep... Idk if it's something about the working directory maybe? I'm running this inside the `api-tests` directory

Wynndow replied:
Yeah same. I'll have a think

Wynndow replied:
Fixed it - prettier does in fact NOT use gitignore. Had to add an ignore file

Joe-Edwards-GDS commented on api-tests/README.md:
Could be worth recommending the Cucumber VSCode plugin (if you're using VSCode?) and how to configure it. I had to update the settings to let it pick up the glue (aka step definitions), as by default it looks for `features/**/*.ts`

Wynndow replied:
Good shout - I'm on intellij but had to do the same.

Joe-Edwards-GDS commented on api-tests/tsconfig.json:
Did this come from somewhere? Looks reasonable (although I wouldn't generally bother including all the config options like this - they'll get out of date when we update TS)

Wynndow replied:
Yeah, running `npx tsc --init` gave it to me.
I thought it was useful to see eveything, but tbh it's too much. I'll trim it out.

Joe-Edwards-GDS commented on api-tests/src/utils/jar-generator.ts:
```diff
@@ -0,0 +1,43 @@
+  journeyType: string,
+): Promise<string> => {
+  const payloadData = JSON.parse(
+    fs.readFileSync(
```
Can we use the async method if possible? If you change your import to `node:fs/promises` then you'll get a promise-based version, which is probably what you want :)

Wynndow replied:
Yep can do - but in this scenario why does it matter? We're going to just wait for it anyway?

Joe-Edwards-GDS replied:
It's a good habit because there is pretty much never a good reason to use synchronous I/O in node.

Although you might think `await asyncCall()` is equivalent to doing it synchronously (and in Java it almost is), in Node this will block the entire event loop - meaning no other code can execute while waiting for it, whereas using async/await does not. In a webserver this is pretty fatal (all requests are blocked!), but even here, we have no idea what other processes cucumber might be trying to run in the background.

Wynndow replied:
Ah, interesting. Thanks

Joe-Edwards-GDS commented on api-tests/src/utils/request-body-generators.ts:
```diff
@@ -0,0 +1,95 @@
+const readJsonFile = (criId: string, scenario: string, jsonType: JsonType) => {
+  return JSON.stringify(
+    JSON.parse(
+      fs.readFileSync(
```
(ditto about async)

Joe-Edwards-GDS commented on api-tests/src/utils/request-body-generators.ts:
```diff
@@ -0,0 +1,95 @@
+    `redirect_uri=${encodeURI(config.ORCHESTRATOR_REDIRECT_URL)}&` +
+    `client_id=${ORCHESTRATOR_CLIENT_ID}&` +
+    `client_assertion_type=urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer&` +
+    `client_assertion=${await createSignedJwt({ sub: ORCHESTRATOR_CLIENT_ID, iss: ORCHESTRATOR_CLIENT_ID })}`
```
You could use `URLSearchParams` here - that would do all the escaping for you :)
```
const params = new URLSearchParams();
params.set("grant_type", "authorization_code")
params.set("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
// etc.
return params.toString();
```
(You might need to prepend a `?` if that's important)

Wynndow replied:
Oh nice.

Joe-Edwards-GDS commented on api-tests/src/utils/request-body-generators.ts:
```diff
@@ -0,0 +1,95 @@
+
+const readJsonFile = (criId: string, scenario: string, jsonType: JsonType) => {
+  return JSON.stringify(
+    JSON.parse(
```
This appears to parse and then immediately stringify - is that needed?

Wynndow replied:
Nope!

Joe-Edwards-GDS commented on api-tests/src/clients/core-back-external-client.ts:
```diff
@@ -0,0 +1,43 @@
+  tokenResponse: TokenResponse,
+): Promise<UserIdentity> => {
+  const response = await fetch(
+    config.CORE_BACK_EXTERNAL_API_URL + `/user-identity`,
```
If you're using backticks here, I think you meant `${config.CORE_BACK_EXTERNAL_API_URL}/user-identity`?

Joe-Edwards-GDS commented on api-tests/src/clients/core-back-internal-client.ts:
If we do anything more complex here, it's probably worth introducing something like `Axios` to help with the API calls - that lets you set up clients etc. with the correct base URL and headers.

Effectively a streamlined version of the core-front implementation https://github.com/govuk-one-login/ipv-core-front/blob/main/src/services/coreBackService.js

Joe-Edwards-GDS commented on api-tests/src/interfaces/journey-engine-response.ts:
This isn't very idiomatic - at the moment you've declared that all responses have all these properties! If we want a single type to capture everything, you should mark them as optional:
```
export interface JourneyEngineResponse {
  page?: string;
  journey?: string;
  cri?: CriResponse;
  client?: ClientResponse;
}
```

However, if you want to be even more precise, you can do something like:
```
export type JourneyEngineResponse =
  | JourneyResponse
  | PageResponse
  | CriResponse
  | ClientResponse;

interface JourneyResponse {
  journey: string;
}

interface PageResponse {
  page: string;
}

interface CriResponse {
  cri: {
    id: string;
    redirectUrl: string;
  };
}

interface ClientResponse {
  client: {
    redirectUrl: string;
  };
}
```

This is a bit more precise, as you can only get each specific combination of properties, but it does mean you need to use a type guard to know which one you actually have, e.g.:
```
const isPageResponse = (
  journeyResponse: JourneyEngineResponse,
): journeyResponse is PageResponse => !!(journeyResponse as PageResponse).page;

const journeyResponse: JourneyEngineResponse = getJourneyResponse();

if (isPageResponse(journeyResponse)) {
  console.log(journeyResponse.page); // happy
  console.log(journeyResponse.cri); // Propert 'cri' does not exist on type 'PageResponse'
}
```

Joe-Edwards-GDS replied:
Worth noting that if we have control over the types, you'd usually prefer to have a discriminator on the type, which avoids the need to write a dedicated type guard.

If we add a `type` property to our responses, then it gets much cleaner:
```
export type JourneyEngineResponse =
  | PageResponse
  | CriResponse;

interface PageResponse {
  type: 'page';
  page: string;
}

interface CriResponse {
  type: 'cri';
  cri: {
    id: string;
    redirectUrl: string;
  };
}

const journeyResponse: JourneyEngineResponse = getJourneyResponse();

if (journeyResponse.type === 'page') {
  console.log(journeyResponse.page); // happy
  console.log(journeyResponse.cri); // Propert 'cri' does not exist on type 'PageResponse'
}
```
Typescript is clever enough to realise that if you check the type is `page`, then it's a `PageResponse` 🎉 

(I wouldn't be against making that change just to make it easier to use in TS)

Wynndow replied:
The way it currently is - why does it work? If I've not marked them as nullable, and we're getting null values (presumably), why hasn't it blown up? I guess because the type system doesn't exist at runtime?

Joe-Edwards-GDS replied:
Yep exactly - the types are all erased at runtime, so only get checked if you write some JS code to check them.

Joe-Edwards-GDS commented on api-tests/src/interfaces/process-cri-callback-request.ts:
```diff
@@ -0,0 +1,6 @@
+  authorizationCode: string | null;
+  credentialIssuerId: string | null;
+  redirectUri: string | null;
+  state: string | null;
```
are these really nullable?

Wynndow replied:
no - as discussed I've removed them as I was already doing a null check on them

Joe-Edwards-GDS commented on api-tests/src/interfaces/process-cri-callback-response.ts:
```diff
@@ -0,0 +1,3 @@
+export interface ProcessCriCallbackResponse {
```
I think it technically returns a `JourneyEngineResponse` - I think if there's an error it'll return a page id rather than a journey step.

Joe-Edwards-GDS commented on api-tests/src/interfaces/redirect-params.ts:
unused?

Wynndow replied:
unused.

Joe-Edwards-GDS commented on api-tests/src/steps/ipv-steps.ts:
```diff
@@ -0,0 +1,95 @@
+  },
+);
+
+Then("I get a client Oauth response", function (this: World): void {
```
(I think I'd prefer OAuth 😄 )

Wynndow replied:
Took me longer than it should have to spot the difference

Joe-Edwards-GDS reviewed (commented):
I might go over this again later, but it's looking pretty good I think!

I'm sure we'll find some tricky cases lurking in the existing test scenarios, but this is pretty much what I expected as a foundation. 🙂

Joe-Edwards-GDS commented on .github/workflows/pre-merge-checks.yml:
```diff
@@ -130,3 +130,34 @@ jobs:
+
+      - name: Install
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.PAT_READ_PACKAGES_SCOPE_CHRIS_WYNNE }}
```
This probably shouldn't have your name on it...!

The reads seem to work for the gradle build using `secrets.GITHUB_TOKEN`, but I guess you tried that. If we need configuration at the data-vocab end, then we can reach out to one of the data architects.

In any case, we probably can't run these tests pre-merge yet, we'll need to set up an environment for them to run against (it's no good running against build if they rely on parallel code changes!)

See https://govukverify.atlassian.net/browse/PYIC-5878

Wynndow replied:
Yeah, i tried with the GITHUB_TOKEN but no dice, unless I did something else wrong. 
But yeah I guess there's no point running against build - will drop this change.

Wynndow committed:
PYIC-5875: Manage env vars with dotenv

Adds dotenv to the project to help with managing the env vars needed to
run.

Wynndow committed:
PYIC-5975: Update README with IDE plugin info

Wynndow committed:
PYIC-5875: Slim down tsconfig.json

Running `npx tsc --init` originall genereated the file with all the
options available commented. This was a bit much.

Wynndow committed:
PYIC-5875: Switch to async file reading

So that we don't block node's event loop. There might be other things
going on in the background. Reading the file synchronously will block
them.

Wynndow committed:
PYIC-5875: Use URLSearchParams to construct token body

It's a lot nicer than cludging together a big string.

Wynndow committed:
PYIC-5875: Use backticks for string munging

It seems a bit nicer than concatenating things together.

Wynndow committed:
PYIC-5875: Refactor journey engine response

This changes it to a type that can take the form of a number of
interfaces. This ensures that we can only get one type of response.

Wynndow committed:
PYIC-5875: Remove | null from process cri callback request

These aren't needed as we're doing a null check before creating an
object with this interface.

Also remove a redundant file.

Wynndow committed:
PYIC-5875: Remove process-cri-callback-response

It could actually return a journey response or a page response in the
case of error. Which we already have interfaces for.

Wynndow committed:
PYIC-5875: Adjust language in feature file

To be less clunky

Wynndow committed:
PYIC-5875: Remove pre-merge API tests

They need to run against an ephemeral env, rather than against build.
Otherwise you'd never be able to get a breaking release out. There's a
ticket for spinning up those new pre-merge envs.

Wynndow committed:
PYIC-5875: Add .prettierignore

Prettier was checking and failing the html report generated by cucumber
(someone should tell cucumber...).

Joe-Edwards-GDS commented on api-tests/.env.template:
```diff
@@ -0,0 +1,17 @@
+CORE_BACK_EXTERNAL_API_URL="https://api.identity.build.account.gov.uk"
+CORE_BACK_PUBLIC_ENCRYPTION_KEY='{"kty":"RSA","e":"AQAB","n":"nel7ibmSTaXWhwEAdqKTiEVcxsYgv6CdXaz90aVN7IorlaCeNj0j06OsA4zdmWEjj21wEZULsxPoZo5N_tsQ7NtOnOkcnDc-g_Nbpt0jelzJSbFRkx3kwXy8YIYKR_myNbiHNTTc7S6GkQRg0N1MPWtzoEKYJs41AN4onrsvUzgpCypWwPy2-ppsaDvms_11YA7A7x3zHj9oKCPJ_uk_0MV3vZAxCxbiPb9ABGWcoGQ5QKGfv40ylBsEdOhE3w-3SAAQIrrHyMRGGiPxcNO161XVL-lOnYt93FgEe16LgpfE22UdENfHnG0UQaTgph1Dm24oqn7qpPTY2DfER ...
+ORCHESTRATOR_REDIRECT_URL="https://orch.stubs.account.gov.uk/callback"
+JAR_SIGNING_KEY='{"kty":"EC","d":"OXt0P05ZsQcK7eYusgIPsqZdaBCIJiW4imwUtnaAthU","crv":"P-256","x":"E9ZzuOoqcVU4pVB9rpmTzezjyOPRlOmPGJHKi8RSlIM","y":"KlTMZthHZUkYz5AleTQ8jff0TJiS3q2OB9L5Fw4xA04"}' # pragma: allowlist secret
```
(I wonder if it's worth calling this `JAR_PUBLIC_SIGNING_KEY` just to stop anyone having a heart attack)

Wynndow replied:
Might be a bit confusing as this a private key - it's the test key we use all over the place and is already in the public domain. Could add a comment to make that obvious?

Joe-Edwards-GDS commented on api-tests/src/config/config.ts:
```diff
@@ -0,0 +1,14 @@
+import "dotenv/config";
```
Do you need to add this to package.json? I get 'could not find package dotenv'

Wynndow replied:
How weird. My cmd history shows I installed with save-dev, must have lost the changes to package.json somehow. Will add it in.

Joe-Edwards-GDS replied:
save yourself some clicks `npm i -D dotenv` :P

Wynndow replied:
If only my poor aching fingers knew about this sooner!

Joe-Edwards-GDS commented on api-tests/src/types/journey-engine-response.ts:
I don't think it's worth separating types and interfaces - I'd probably just put them all under `types` (the distinction isn't important, the point is that this is where all the type definitions are).

I'd also combine `journey-engine-response.ts` and `journey-engine-responses.ts` into a single file - they naturally belong together 😄 

(You could also put all the 'internal API types' in a single file, should you wish)

Joe-Edwards-GDS commented on api-tests/src/utils/type-guards.ts:
This is probably too generic a filename - I'd make it `journey-response-helpers.ts` or something.

(Alternatively, you could bundle them in with the type definitions)

Joe-Edwards-GDS commented on api-tests/src/interfaces/journey-engine-responses.ts:
Technically I think there's also `ErrorResponse` - it's the same as `PageResponse` but with an added `statusCode`.

(Might be better to just add an optional `statusCode` onto `PageResponse`!)

Wynndow replied:
Ah that's true. It does feel like a special case of a Page response, so I'll add an optional

Joe-Edwards-GDS commented on .pre-commit-config.yaml:
```diff
@@ -32,3 +32,11 @@ repos:
+        name: Lint API tests
+        entry: bash -c 'cd api-tests && npm run lint'
+        language: system
+        pass_filenames: false
```
Is it possible to filter this so it only runs if you actually change something in `api-tests`?

Wynndow replied:
Turns out yes!

Wynndow committed:
PYIC-5875: Reorganise types

This moves all the interface and type definitions into the types dir. It
also groups them in files bases around the client interactions.

Wynndow committed:
PYIC-5875: Only run pre-commit check on api-tests

This filters the npm-run-lint pre-commit hook if there is a change to
the api-tests directory, rather than any change.

Wynndow committed:
PYIC-5875: Include EVCS access token in JAR

We're going to need this when we start turing on EVCS writes and reads.
Handily the dev and build EVCS stub is configured to use the same
signing key as the orch stubs JAR requests, so we can reuse the already
defined signing key.

Joe-Edwards-GDS commented on .gitignore:
```diff
@@ -53,3 +53,8 @@ local-running/bin/
+api-tests/cucumber-report.html
+api-tests/.npmrc
+api-tests/.env
+api-tests/src/**/*.js
```
(This addition looks a little suspicious...)

Wynndow replied:
I added this as I ran `npx tsc` to try compiling stuff with typecript directly, rather than using ts-node. And it generated all the js files, which we don't want. Is there a better way to deal with this?

Joe-Edwards-GDS replied:
You can add `noEmit: true` to your tsconfig.json to prevent it from emitting JS files - then `tsc` will just do type-checking.

Joe-Edwards-GDS reviewed (commented):
Looks good to me - we can always revise things as we develop more test cases etc.

Let's wait until someone from QA has had a look before merging.

Wynndow committed:
PYIC-5875: Set noEmit to true in tsconfig

We're using ts-node which compiles TS into JS on the fly when the tests
run. We're not doing any compilation step - so we don't want to emit any
JS files. This means we can also remove JS files from the gitignore.

Wynndow committed:
PYIC-5875: Add api-tests to dependabot.yaml

So we can stay fresh
